src/main/java/com/uber/cadence/samples/hello/HelloAsyncWorkflowWithSignalCompletion.java

```diff
@@ -0,0 +1,172 @@
+ * local instance of Cadence server to be running. 1. Start Workflow (asynchronous) 2. Wait 3.
+ * Signal-1 => Do Activity-1 4. Wait 5. Signal-2 => Do Activity-2 6. Query Workflow 7. End Workflow
+ */
+public class HelloAsyncWorkflowWithSignalCompletion {
+
+  static final String TASK_LIST = "HelloAsyncWorkflowWithSignalCompletion";
+
+  /** Workflow interface must have a method annotated with @WorkflowMethod. */
+  public interface GreetingWorkflow {
+    /**
+     * @return list of greeting strings that were received through the waitForNameMethod. This
+     *     method will block until the number of greetings specified are received.
+     */
+    @WorkflowMethod
+    List<String> getGreetings();
+
+    /** Signal1 - Receives signal1 through an external signal. */
+    @SignalMethod
+    void waitForSignal1(String signal1);
+
+    /** Signal2 - Receives signal2 through an external signal. */
+    @SignalMethod
+    void waitForSignal2(String signal2);
+
+    @QueryMethod
+    String queryResults();
+  }
+
+  /** Activity interface is just a POJI. * */
+  public interface GreetingActivity1 {
+    @ActivityMethod(scheduleToCloseTimeoutSeconds = 10)
+    String composeGreeting1(String greeting, String name);
+  }
+
+  /** Activity interface is just a POJI. * */
+  public interface GreetingActivity2 {
+    @ActivityMethod(scheduleToCloseTimeoutSeconds = 10)
+    String composeGreeting2(String greeting, String name);
+  }
+
+  /** GreetingWorkflow implementation that returns a greetings as per the designed workflow */
+  public static class GreetingWorkflowImpl implements GreetingWorkflow {
+
+    List<String> messageQueue = new ArrayList<>(2);
+    //    boolean exit = false;
+    CompletablePromise<String> signal1 = Workflow.newPromise();
+    CompletablePromise<String> signal2 = Workflow.newPromise();
+
+    /**
+     * Activity stub implements activity interface and proxies calls to it to Cadence activity
+     * invocations. Because activities are reentrant, only a single stub can be used for multiple
+     * activity invocations.
+     */
+    private final GreetingActivity1 activity1 = Workflow.newActivityStub(GreetingActivity1.class);
+
+    private final GreetingActivity2 activity2 = Workflow.newActivityStub(GreetingActivity2.class);
+
+    @Override
+    public List<String> getGreetings() {
+      List<String> receivedMessages = new ArrayList<>(5);
+
+      System.out.println("Awaiting Signal1");
+      if (!signal1.get().isEmpty()) {
+        String activity1Result = activity1.composeGreeting1("Hello", "Activity-1");
+        receivedMessages.add(activity1Result);
+      }
+
+      System.out.println("Awaiting Signal2");
+      if (!signal2.get().isEmpty()) {
+        String activity2Result = activity2.composeGreeting2("Hello", "Activity-2");
+        receivedMessages.add(activity2Result);
+      }
+
+      messageQueue.addAll(receivedMessages);
+      return receivedMessages;
```

Comment:
I would return just activity1Result + activity2Result, may be with some separator.